core/server/master/src/main/java/alluxio/master/metastore/caching/Cache.java

```diff
@@ -0,0 +1,410 @@
+        onCacheUpdate(key, value.get());
+        Entry newEntry = new Entry(key, value.get());
+        newEntry.mDirty = false;
+        return newEntry;
+      }
+      return null;
+    });
+    if (entry == null || entry.mValue == null) {
+      return Optional.empty();
+    }
+    entry.mReferenced = true;
+    wakeEvictionThreadIfNecessary();
+    return Optional.of(entry.mValue);
+  }
+
+  /**
+   * Writes a key/value pair to the cache.
+   *
+   * @param key the key
+   * @param value the value
+   */
+  public void put(K key, V value) {
+    mMap.compute(key, (k, entry) -> {
+      onNew(key, value);
+      if (entry == null && cacheIsFull()) {
+        writeToBackingStore(key, value);
+        return null;
+      }
+      if (entry == null || entry.mValue == null) {
+        onCacheUpdate(key, value);
+        return new Entry(key, value);
+      }
+      entry.mValue = value;
+      entry.mReferenced = true;
+      entry.mDirty = true;
+      return entry;
+    });
+    wakeEvictionThreadIfNecessary();
+  }
+
+  /**
+   * Removes a key from the cache.
+   *
+   * The key is not immediately removed from the backing store. Instead, we set the entry's value to
+   * null to indicate to the eviction thread that to evict the entry, it must first remove the key
+   * from the backing store. However, if the cache is full we must synchronously write to the
+   * backing store instead.
+   *
+   * @param key the key to remove
+   */
+  public void remove(K key) {
+    // Set the entry so that it will be removed from the backing store when it is encountered by
+    // the eviction thread.
+    mMap.compute(key, (k, entry) -> {
+      if (entry == null && cacheIsFull()) {
+        removeFromBackingStore(k);
+        return null;
+      }
+      onCacheUpdate(key, null);
+      onRemove(key);
+      if (entry == null) {
+        entry = new Entry(key, null);
+      } else {
+        entry.mValue = null;
+      }
+      entry.mReferenced = false;
+      entry.mDirty = true;
+      return entry;
+    });
+    wakeEvictionThreadIfNecessary();
+  }
+
+  /**
+   * Clears all entries from the map. This is not threadsafe, and requires external synchronization
+   * to prevent concurrent modifications to the cache.
+   */
+  public void clear() {
+    mMap.forEach((key, value) -> {
+      onCacheUpdate(key, value.mValue);
+      onEvict(key, null);
+    });
+    mMap.clear();
+  }
+
+  private boolean overLowWaterMark() {
+    return mMap.size() >= mLowWaterMark;
+  }
+
+  private boolean overHighWaterMark() {
+    return mMap.size() >= mHighWaterMark;
+  }
+
+  private boolean cacheIsFull() {
+    return mMap.size() >= mMaxSize;
+  }
+
+  private void wakeEvictionThreadIfNecessary() {
+    if (mEvictionThread.mIsSleeping && mMap.size() >= mHighWaterMark) {
+      kickEvictionThread();
+    }
+  }
+
+  private void kickEvictionThread() {
+    synchronized (mEvictionThread) {
+      if (mEvictionThread.getState() == State.NEW) {
```

Comment:
it's intentional because starting a thread from a constructor is considered harmful https://stackoverflow.com/questions/84285/calling-thread-start-within-its-own-constructor

added a comment in the constructor